rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,854 @@
+        let av = AccessVector::from_value(&(*self.0).vec_pages);
+
+        // Which block (rank relative to this region)?
+        let block_rank = offset / meta_data::size::BLOCK_IN_BYTES;
+
+        // Where in that block?
+        let intra_block_index = offset % meta_data::size::BLOCK_IN_BYTES;
+
+        // Where is that block located in stable memory (global rank)?
+        let block_id = av.get_ith_block_id(block_rank as u32);
+
+        // address of the byte to load from stable memory:
+        let offset = meta_data::offset::BLOCK_ZERO
+            + block_id.0 as u64 * meta_data::size::BLOCK_IN_BYTES
+            + intra_block_index;
+        (
+            offset,
+            block_id,
+            meta_data::size::BLOCK_IN_BYTES - intra_block_index,
+        )
+    }
+
+    // compute the absolute offset, begin block, remaining length of
+    // begin block, and end block for a relative offset and length.
+    //
+    // NB: BlockIds can be used to do case analysis (same or diff
+    // block?) when planning successive reads/writes.
+    pub unsafe fn relative_into_absolute_span(
+        &self,
+        offset: u64,
+        len: u64,
+    ) -> (u64, BlockId, u64, BlockId) {
+        let (off, b1, b1_len) = self.relative_into_absolute_info(offset);
+        if len == 0 {
+            return (off, b1.clone(), b1_len, b1);
+        };
+        let final_offset = offset + len - 1;
+        let (_, b2, _) = self.relative_into_absolute_info(final_offset);
+        (off, b1, b1_len, b2)
+    }
+}
+
+// Mutable meta data stored in stable memory header (See motoko/design/StableRegions.md)
+mod meta_data {
+
+    /// Maximum number of entities.
+    pub mod max {
+        pub const BLOCKS: u16 = 32768;
+        pub const REGIONS: u16 = 32767;
+    }
+
+    /// Sizes of table entries, and tables.
+    pub mod size {
+        // Region table entry:
+        // - 8 bytes for the size.
+        // - 4 bytes for future use (including GC visit-marking).
+        pub const REGION_TABLE_ENTRY: u16 = 12;
+
+        pub const BLOCK_REGION_TABLE_ENTRY: u16 = 4;
+
+        pub const BLOCK_REGION_TABLE: u64 =
+            super::max::BLOCKS as u64 * BLOCK_REGION_TABLE_ENTRY as u64;
+
+        pub const PAGES_IN_BLOCK: u32 = 128;
+        pub const PAGE_IN_BYTES: u64 = 1 << 16;
+        pub const BLOCK_IN_BYTES: u64 = PAGE_IN_BYTES * (PAGES_IN_BLOCK as u64);
+
+        // Static memory footprint, ignoring any dynamically-allocated pages.
+        pub const STATIC_MEM_IN_PAGES: u64 = PAGES_IN_BLOCK as u64; /* One block for meta data, plus future use TBD. */
+
+        pub unsafe fn required_pages() -> u64 {
+            STATIC_MEM_IN_PAGES
+                + (super::total_allocated_blocks::get() as u64) * (PAGES_IN_BLOCK as u64)
+        }
+    }
+
+    /// Offsets into stable memory for statically-sized fields and tables.
+    pub mod offset {
+        pub const TOTAL_ALLOCATED_BLOCKS: u64 = 0;
+
+        pub const TOTAL_ALLOCATED_REGIONS: u64 = 2;
```

Comment:
Perhaps, a comment could be helpful: One region is reserved for `region0`, the other as reclamation pool?